File: apps/ios/Pebbles/Theme/PebblesList.swift
import SwiftUI

// MARK: - Row position

/// Where a row sits inside its Section, used to mask the border overlay's
/// corner radii. `.only` is the default for single-row sections.
enum PebblesListRowPosition {
    case only
    case top
    case middle
    case bottom
}

func pebblesRowPosition(index: Int, count: Int) -> PebblesListRowPosition {
    if count <= 1 { return .only }
    if index == 0 { return .top }
    if index == count - 1 { return .bottom }
    return .middle
}

// MARK: - List/Form chrome

/// Applied to `List` or `Form`: hides the native grouped background,
/// recolors row separators to `system.muted`, and sets `Spacing.lg`
/// between sections so the bordered groups breathe consistently.
extension View {
    func pebblesList() -> some View {
        modifier(PebblesListModifier())
    }
}

private struct PebblesListModifier: ViewModifier {
    func body(content: Content) -> some View {
        content
            .scrollContentBackground(.hidden)
            .listRowSeparatorTint(Color.system.muted)
            .listSectionSpacing(Spacing.lg)
    }
}

// MARK: - Row chrome

/// Applied to each row inside a Section: replaces the native row background
/// with a `system.muted` stroke that draws only the edges this row owns —
/// top/sides for the first row, sides only for middle rows, bottom/sides for
/// the last. Combined with the list-level `listRowSeparatorTint` (also
/// `system.muted`), the horizontal dividers between rows complete the card.
extension View {
    func pebblesListRow(position: PebblesListRowPosition = .only) -> some View {
        modifier(PebblesListRowModifier(position: position))
    }
}

private struct PebblesListRowModifier: ViewModifier {
    let position: PebblesListRowPosition

    func body(content: Content) -> some View {
        content
            .listRowBackground(borderBackground)
            .listRowSeparator(.hidden)
    }

    @ViewBuilder
    private var borderBackground: some View {
        let radius = Spacing.lg
        switch position {
        case .only:
            RoundedRectangle(cornerRadius: radius)
                .stroke(Color.system.muted, lineWidth: 1)
        case .top:
            UnevenRoundedRectangle(
                cornerRadii: RectangleCornerRadii(
                    topLeading: radius,
                    bottomLeading: 0,
                    bottomTrailing: 0,
                    topTrailing: radius
                )
            )
            .stroke(Color.system.muted, lineWidth: 1)
        case .middle:
            Rectangle()
                .stroke(Color.system.muted, lineWidth: 1)
        case .bottom:
            UnevenRoundedRectangle(
                cornerRadii: RectangleCornerRadii(
                    topLeading: 0,
                    bottomLeading: radius,
                    bottomTrailing: radius,
                    topTrailing: 0
                )
            )
            .stroke(Color.system.muted, lineWidth: 1)
        }
    }
}

// MARK: - Section header

/// Section header typography matching profile cards (Stats, Collections):
/// `.pebblesFont(.cardHeading)` (SF Compact Rounded 15 semibold, uppercase,
/// 10% tracking) in `system.secondary`.
extension Text {
    func pebblesSectionHeader() -> some View {
        self
            .pebblesFont(.cardHeading)
            .foregroundStyle(Color.system.secondary)
    }
}

// MARK: - Preview

#Preview("PebblesList chrome") {
    Form {
        Section {
            Text("Single row")
                .pebblesListRow(position: .only)
        } header: {
            Text("Single").pebblesSectionHeader()
        }

        Section {
            Text("Top row").pebblesListRow(position: .top)
            Text("Middle row").pebblesListRow(position: .middle)
            Text("Bottom row").pebblesListRow(position: .bottom)
        } header: {
            Text("Multi-row").pebblesSectionHeader()
        }
    }
    .pebblesList()
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/theme/PebblesList.kt
package app.pbbls.android.theme

import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.unit.dp

/**
 * Where a row sits inside its section, used to pick which corners of the
 * 1dp `system.muted` border round. `ONLY` is the default for single-row
 * sections. Mirrors iOS `Theme/PebblesList.swift`.
 */
enum class PebblesListRowPosition {
    ONLY,
    TOP,
    MIDDLE,
    BOTTOM,
}

/** Pure position rule — ports `pebblesRowPosition(index:count:)` (JVM-tested). */
fun pebblesRowPosition(
    index: Int,
    count: Int,
): PebblesListRowPosition =
    when {
        count <= 1 -> PebblesListRowPosition.ONLY
        index == 0 -> PebblesListRowPosition.TOP
        index == count - 1 -> PebblesListRowPosition.BOTTOM
        else -> PebblesListRowPosition.MIDDLE
    }

/**
 * Row chrome — the `pebblesListRow(position:)` analog: a 1dp `system.muted`
 * border whose corner radii (Spacing.lg) round only the edges this row owns.
 * [PebblesListSection] overlaps adjacent rows by the border width so the
 * shared edge renders as a single divider, completing the bordered card.
 */
fun Modifier.pebblesListRow(position: PebblesListRowPosition = PebblesListRowPosition.ONLY): Modifier =
    composed {
        val radius = PebblesTheme.spacing.lg
        val shape =
            when (position) {
                PebblesListRowPosition.ONLY -> RoundedCornerShape(radius)
                PebblesListRowPosition.TOP -> RoundedCornerShape(topStart = radius, topEnd = radius)
                PebblesListRowPosition.MIDDLE -> RoundedCornerShape(0.dp)
                PebblesListRowPosition.BOTTOM -> RoundedCornerShape(bottomStart = radius, bottomEnd = radius)
            }
        border(1.dp, PebblesTheme.colors.system.muted, shape)
    }

/**
 * A bordered section of rows — the `List`/`Form` + `pebblesListRow` idiom as
 * one composable, since Compose has no grouped-list container. Each row is
 * wrapped in the position-aware border and given the standard row insets;
 * rows overlap by 1dp so adjacent borders coincide into a single divider.
 *
 * [rows] is an explicit list (not a free content block) because the border
 * needs each row's index/count — the same reason the iOS API takes an
 * explicit `position:`.
 */
@Composable
fun PebblesListSection(
    rows: List<@Composable () -> Unit>,
    modifier: Modifier = Modifier,
    header: String? = null,
) {
    Column(modifier = modifier.fillMaxWidth()) {
        if (header != null) {
            PebblesSectionHeader(
                text = header,
                modifier = Modifier.padding(bottom = PebblesTheme.spacing.sm),
            )
        }
        Column(verticalArrangement = Arrangement.spacedBy((-1).dp)) {
            rows.forEachIndexed { index, row ->
                Box(
                    modifier =
                        Modifier
                            .fillMaxWidth()
                            .pebblesListRow(pebblesRowPosition(index, rows.size))
                            .padding(horizontal = PebblesTheme.spacing.lg, vertical = PebblesTheme.spacing.md),
                ) {
                    row()
                }
            }
        }
    }
}

/**
 * Section header typography matching profile cards — the
 * `pebblesSectionHeader()` analog: `cardHeading` token (uppercase via
 * [PebblesText]) in `system.secondary`.
 */
@Composable
fun PebblesSectionHeader(
    text: String,
    modifier: Modifier = Modifier,
) {
    PebblesText(
        text = text,
        style = PebblesTypography.cardHeading,
        color = PebblesTheme.colors.system.secondary,
        modifier = modifier,
    )
}
